fastestimator/trace/io/model_eval.py

```diff
@@ -0,0 +1,446 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import inspect
+import json
+import locale
+import os
+import re
+import shutil
+from datetime import datetime
+from time import time
+from typing import Callable, List, Union
+
+import numpy as np
+import pydot
+from pylatex import Command, Document, Figure, Hyperref, Itemize, Label, LongTable, Marker, MultiColumn, NoEscape, \
+    Package, Section, Subsection, Subsubsection, Table, Tabular, Tabularx, TextColor, escape_latex
+from pylatex.utils import bold, dumps_list
+
+import fastestimator as fe
+from fastestimator.summary.logs.log_plot import visualize_logs
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+from fastestimator.util.latex_util import PyContainer, TabularCell, WrapText
+from fastestimator.util.traceability_util import get_environment, traceable
+from fastestimator.util.util import to_list, to_number
+
+
+@traceable()
+class TestCase():
+    """This class defines the test case that TestReport trace will take to perform auto-testing.
+
+    Args:
+        description: A test description.
+        criteria: Function to perform the test that return True when test passes and False when test fails. Input
+            variable name will be used as input keys to futher derive input value.
+        aggregate: If True, this test will be treated as per-instance type of which test criteria will be examined at
+            batch_end. If False, this test is aggregate type and its criteria will be examined at epoch_end.
+        fail_threshold: Thershold of failing sample number to judge sample-case test as failed or passed. If failing
+            number is above this value, then the test fails; otherwise it passes. It only has effect when sample_wise
+            equal to true.
+    """
+    def __init__(self, description: str, criteria: Callable, aggregate: bool = True, fail_threshold: int = 0) -> None:
+        self.description = description
+        self.criteria = criteria
+        self.criteria_inputs = inspect.signature(criteria).parameters.keys()
+        self.aggregate = aggregate
+        if not self.aggregate:
+            self.fail_threshold = fail_threshold
+
+    def clean_result(self) -> None:
+        if not self.aggregate:
+            self.result = []
+            self.fail_id = []
+
+
+@traceable()
+class ModelEval(Trace):
+    """Automate testing and report generation.
+
+    Args:
+        test_cases: List of TestCase object.
```

Comment:
object**s**, except it's also not always a list based on your typing below

Reply:
fixed